Async web3

fselmo opened the pull request:
### What was wrong?

Related to Issue #1416 

- Separates ``AsyncWeb3`` from ``Web3`` in an effort to make async a first class citizen of the library. Ideally we refactor the structure of the files so they are organized as such for developers. I feel like there are enough changes here where re-organization should be a separate PR, which will be quite large as well.
- Create ``AsyncMiddleware`` type that expects ``AsyncWeb3`` rather than ``Web3``. Account for this change down the tree through async middlewares and middleware onion.
- Remove all cases of ``type: ignore`` that were fixed by the tighter typing for async web3 and related modules.

Things to note:

- ``AsyncContractCaller`` inherits from ``BaseContractCaller`` which used the synchronous method ``parse_block_identifier`` in its ``__init__()`` method. This leads me to think we weren't testing all cases for initialization of the contract caller. This came up as an issue with typing and was addressed using the simpler ``parse_block_identifier_no_extra_call()`` method which technically shouldn't break anything since the previous structure would return a coroutine and also not work. This may need to be addressed in the future if we want to try to account for all elif cases that ``ContractCaller`` accounts for.

### How was it fixed?

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)
- [x] ~~Update documentation + document new async flow~~ (separate PR, see #2821)

#### Cute Animal Picture

![20230207_091433](https://user-images.githubusercontent.com/3532824/219521431-6da073b5-54d3-42e0-959c-9c71bd493bc6.jpg)

fselmo committed:
Initial refactoring for async typing support

- Create ``AsyncWeb3`` since ambiguity when instantiating modules with ``Union`` types leads to typing issues. Python's static typing forces us to make the async library a first class citizen.
- Refactor appropriately down the tree, accounting for the newly expected ``AsyncWeb3`` class in things like middlewares and modules, including ``ens``.
- Be more explicit within modules which type of ``w3`` is expected, ``Web3`` or ``AsyncWeb3``, as in the ``EthPM`` module, for example.
- Resolve some ``type: ignore`` cases along the way as some of the typing gets tightened.

fselmo committed:
Removed all remaining cases of type: ignore

fselmo committed:
Fix async issue with parsing block identifier

- ``BaseContractCaller`` used the sync version of ``parse_block_identifier`` which would render any calls inside it useless for ``AsyncContractCaller`` which inherits from it. I'm assuming we don't test all cases here. This commit splits the bulk of that logic out to the sync and async __init__ methods for contract caller and calls a more appropriate method for parsing the block identifier that doesn't involve an extra call. This reduces some of the intended functionality, but that wasn't working to begin with.

This will need to be revised at some point to see if we can account for all desired cases within the async contract caller __init__.

fselmo committed:
Fix issues with rebasing and update description for simpler block_id method

kclowes reviewed (commented):
Hooray! :rocket: I left some comments around adding a new `TypeVar` instead of `Union["ContractFunction", "AsyncContractFunction"]`, that I think can help avoid some of the `cast`ing we're doing in the `contract`/`async_contract` files. It was sort of hard to figure out where to put the comments, so lmk if they don't make sense and we can hop on a call!

fselmo committed:
Address comments on PR #2819; add newsfragment

- Improve typing by using `TypeVar` over `Union` in certain places to prevent casting
- Update links for geth namespaced api docs

fselmo reviewed (commented):
@kclowes I addressed your comments in the latest commit. ~~It looks like some typing broke for `filter_builder.args` when I added the `TypeVar` suggestion. I ignored those for now. It's not immediately obvious to me where that broke.~~ [nevermind, fixed 🎉]

pacrob reviewed (commented):
I left a couple comments, but overall lookin' good! I love all those `# type: ignore`s going away.

kclowes reviewed (approved):
LGTM! 🚢 ! Just had a tiny nit.
